My code is not passing tests and I'm stuck on the listIterator

TBarrett553 opened the pull request:
I'm working on the poorly made conditionals, right now. I apologize this should have been finished last week. I would really appreciate your input. My main problem is that the listIterator is infinite and it only works if I put a break at the end.

TBarrett553 committed:
My code is not passing tests and I'm stuck on the listIterator

TBarrett553 committed:
Merge branch 'week-4' into week-5

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+
+    // Check that there are booked meetings before sorting it and running the available function
+    if (bookTimes.isEmpty()) {
+            availTimes.add(bookTimes.get(0).WHOLE_DAY);
```
Hmm, how does this work? Won't bookTimes.get(0) throw if bookTimes.isEmpty()?

flying-bus replied:
If you're just trying to access the static variable, then I think you can just say: TimeRange.WHOLE_DAY.

TBarrett553 replied:
There is a test that requires that the whole day should be returned if the day is open. In other words, if the bookTimes is empty I wanted to skip the whole list iterator function. There is probably a more elegant way to do this.

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+        for (Iterator<String> attendIterator = requestAttendees.iterator(); attendIterator.hasNext();){
+            String attendee = attendIterator.next();
+           
+            if(event.getAttendees().contains(attendee) && !bookTimes.contains(event.getWhen())){
```
If you used a HashSet for bookTimes, then you wouldn't have to check bookTimes.contains(event.getWhen()). And it would be more efficient.

https://docs.oracle.com/javase/7/docs/api/java/util/HashSet.html

TBarrett553 replied:
The main reason I used an ArrayList was so I could sort it by time. But I initially made a HashSet to go through the for loop and then converted it to a list so it can be sorted.

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+    }
+
+    else {
+        Collections.sort(bookTimes, bookTimes.get(0).ORDER_BY_START);
```
Static variable just use the classname:
TimeRange.ORDER_BY_START

TBarrett553 replied:
Thank you and makes a lot of sense!

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+
+  public Collection<TimeRange> availableTimes(ArrayList<TimeRange> bookTimes, int length){
+    Collection<TimeRange> availTimes = new HashSet<>();
+    if(length > bookTimes.get(0).END_OF_DAY) {
```
I'm not sure what this means - can you add a comment explaining what it does?

If the length of the event is more than a day, then there are no available times?

TBarrett553 replied:
This was in consideration of a test (too long of a duration). The test says that the code should return an empty list if the requested duration is longer than a whole day.

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+                if(!bookedTime.overlaps(prevBookedTime) && (prevBookedTime.end() + length <= bookedTime.start())) {
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.start(), false));
+                }
+
```
A good way to format if/else is like this:

if (cond) {

} else if (cond) {

} else {

}

That way the formatting gives an extra clue that you're looking at an else/if condition.

TBarrett553 replied:
I tried this formatting in my recent commits. Was it done correctly?

flying-bus replied:
The formatting LGTM (Looks Good To Me)

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(bookIterator.hasPrevious()) {
+                TimeRange prevBookedTime = bookIterator.previous();
```
Ah ha, I think I found your bug.

When you call "previous" it actually moves the iterator backwards! So on Line 77, you move the iterator forwards, and then Line 80 moves it backwards again, so you never actually manage to advance the iterator.

If you want to actually store the previous TimeRange, then this might work:

```
TimeRange prevBookedTime = null;
for (ListIIterator<TimeRange> bookIterator = /* like before */) {
  TimeRange bookedTime = bookIterator.next();
  if (prevBookedTime != null) {  // instead of bookIterator.hasPrevious()
     if (!bookedTime.overlaps(prevBookedTime) /* ... like before ... (/ ) {
     }
  }
 
  // Store bookedTime in prevBookedTime for the next iteration through the loop.
  prevBookedTime = bookedTime;
}
```

https://docs.oracle.com/javase/7/docs/api/java/util/ListIterator.html#previous()

TBarrett553 replied:
Wow thank you! When I kept looking into this error I couldn't find an example when .previous() and .next() were used in the same iterator instance. But this makes sense, pointers are a tricky thing I'm still learning about.

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,96 @@
+
+                else if(bookedTime.overlaps(prevBookedTime)) {
+                    if(prevBookedTime.contains(bookedTime)){
+                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.END_OF_DAY, true));
```
Use TimeRange.END_OF_DAY and TimeRange.START_OF_DAY here and below.

TBarrett553 replied:
Got it!

freeboson commented:
I'll take a look after you make the changes @flying-bus requested. Ping me here or in dynamite when you're ready.

TBarrett553 committed:
Fixed previous Bug

I also made the Time Conflicts loop iterate through a HashSet. However, it has to be changed to a list so it can be sorted.

TBarrett553 committed:
Added 5 Extra Tests

As a part of the assignment, we had to make our own tests. The last 5 are the ones I made so if you have suggestions on how to make better tests please do so! Additionally, there are 3 tests that will not pass because they consider Optional Attendees. If there was extra time I could have added the element to my code (I don't have that time).

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,97 @@
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // Initiatizes return data structure and set duration parameter
+    //Collection<TimeRange> availTimes = new HashSet<>();
```
rm

TBarrett553 replied:
Now removed

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,97 @@
+        return availTimes;
+
+    // If there are no Time Conflicts the whole day should be returned as available
+    } else if (bookTimes.isEmpty()) {
```
This doesn't need an else

TBarrett553 replied:
So basically I should make it a separate if statement?

freeboson replied:
Yes, because there's an early return (good — prefer early return over nested ifs)

```
if (cond1) {
  return ___;
}
if (cond2) {
  return ___;
}
```

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,97 @@
+        availTimes.add(TimeRange.WHOLE_DAY);
+        return availTimes;
+
+    } else {
```
This shouldn't be an else

TBarrett553 replied:
Now I see that it's unnecessary

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,97 @@
+        for(ListIterator<TimeRange> bookIterator = bookTimes.listIterator(); bookIterator.hasNext();) {
+            TimeRange bookedTime = bookIterator.next();
+            
+            if(prevBookedTime != null) { 
```
Need a comment before this with what you're doing here. E.g.,

// If we've already found a booked time, extend it

(if that's what the behavior is -- i can't really tell)

TBarrett553 replied:
Yeah that was a mess I was able to clean it up and I placed new comments with the refigured algorithm.

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,91 +19,92 @@
-    return bookTimes;
+
+    // The times have to be returned as a list to be sorted by start time
+    ArrayList<TimeRange> bookedTimes = new ArrayList<TimeRange>(bookTimes) ; 
```
The next two lines may be simpler to do:

return new ArrayList<TimeRange>(bookTimes);

TBarrett553 replied:
Great thank you for that simplification!

flying-bus commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -19,91 +19,92 @@
-                    if(prevBookedTime.contains(bookedTime)){
-                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.end(), bookedTime.END_OF_DAY, true));
-                    availTimes.add(bookedTime.fromStartEnd(prevBookedTime.START_OF_DAY, bookedTime.END_OF_DAY, true));
+                    if(!bookIterator.hasNext() && bookedTime.end() + length <= TimeRange.END_OF_DAY) {
```
There are a lot of nested if-statements here, which makes it sort of hard to follow. Some of them look very similar? (Like here and Line 92).

Is it possible to simplify the conditions a bit? Or if not, can the repeated code be put into a helper function? Or at the very least, add some comments which state what the conditions are.

freeboson replied:
suggestion from 1:1 preprocess `bookTimes` to be disjoint/deduped.

TBarrett553 replied:
I was able to implement the disjoint method.

flying-bus commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
+
+  @Test
+  public void optionalAttendeeConsidered() {
+      // Have each person have different events. We should see three options because each person has
```
s/three/two/

(this is common notation for replace "three" with "two" based syntax for the tool "sed").

TBarrett553 replied:
Got it! Very interesting

TBarrett553 committed:
Removed else conditional

TBarrett553 committed:
isEmpty() conditional is an if statement

flying-bus commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
+  @Test
+  public void optionalAttendeeConsidered() {
+      // Have each person have different events. We should see three options because each person has
+    // split the restricted times. There is an optional attendee and their event is considered.
```
The previous test doesn't consider the optional attendee, but this test does. I don't obviously see where that option is set (is it a flag? or something else that triggers it?).

TBarrett553 replied:
It seems that the optional attendee's schedule is considered if their event has the same TimeRange as an available meeting time. It's kind of an odd requirement but that's why the optional attendee is considered in the test.

flying-bus replied:
I think I just don't see where in the code the optional attendees are considered at all.

TBarrett553 replied:
I don't have the optional feature in my code because the walk through stated that if there is extra time I could add code for the optional tests. Unfortunately I was behind, so I couldn't prioritize this section.

flying-bus replied:
ok, sgtm.

flying-bus commented on walkthroughs/week-5-tdd/project/src/test/java/com/google/sps/FindMeetingQueryTest.java:
```diff
@@ -270,6 +272,151 @@ public void notEnoughRoom() {
+
+  @Test
+  public void optionalAttendeesWithoutGaps() {
+    // There are two optional meeting attendees. They have several gaps in their schedule that 
```
Is this comment accurate?

TBarrett553 committed:
Added Merge Method

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,115 @@
+    return new ArrayList(bookTimes);
+  }
+
+  public ArrayList<TimeRange> collapseTimeConflicts (ArrayList<TimeRange> bookTimes){
```
This method both mutates its argument, and returns it. Consider making it either:

```
public List<TimeRange> collapseTimeConflicts(ImmutableList<TimeRange> bookTimes) {
  // ...
  return value;
}
```

or:

```
public void collapseTimeConflicts(List<TimeRange> bookTimes) {
  // mutate bookTimes directly
}
```

TBarrett553 replied:
So essentially, if I mutate the list I don't need to return it?

freeboson commented on walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java:
```diff
@@ -14,10 +14,115 @@
+                if(prevBookedTime.contains(bookedTime)){
+                    bookTimes.remove(bookedTime); 
+
+                // For overlapping events create a new Time Range with the earliest start time and latest end time
```
It's worth noting that the case where bookedTime.end() is earlier than prevBookedTime.end() is handled by the `.contains()` case above.

TBarrett553 committed:
Updated Comment about Contained Function

TBarrett553 committed:
From Three to Two

TBarrett553 committed:
Fixed No Gaps Comment

TBarrett553 committed:
Made CollapseTimeConflicts a Void Function

The parameter is an immutable list and a new list is no longer returned
